python_modules/dagster/dagster/core/execution/with_resources.py

```diff
@@ -1,28 +1,95 @@
-from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, TypeVar, cast
+from typing import Any, Iterable, List, Mapping, Optional, Sequence, TypeVar, cast
 
 from dagster import _check as check
 from dagster.utils import merge_dicts
 
+from ...config import Shape
 from ..definitions import ResourceDefinition
 from ..definitions.resource_requirement import ResourceAddable
+from ..errors import DagsterInvalidConfigError, DagsterInvalidInvocationError
 
 T = TypeVar("T", bound=ResourceAddable)
 
 
 def with_resources(
     definitions: Iterable[T],
     resource_defs: Mapping[str, ResourceDefinition],
-    config: Optional[Dict[str, Any]] = None,
+    resource_config_by_key: Optional[Mapping[str, Any]] = None,
 ) -> Sequence[T]:
+    """Function that adds dagster resources to copies of resource-requiring dagster definitions.
```

Comment:
We don't typically say "Function" in function docstring. Can just start with "adds".